### PR TITLE
Migrate CLI Kerberos authentication

### PR DESCRIPTION
**Pull request by Jessie212**
## Description

> Is this change a fix, improvement, new feature, refactoring, or other?

Improvement.

> Is this a change to the core query engine, a connector, client library, or the SPI interfaces? (be specific)

Changes made to documentation only.

> How would you describe this change to a non-technical end user or system administrator?

-Created a Kerberos authentication section in the Trino CLI page, migrated content from the CLI Kerberos authentication page to the new Trino CLI page, and deleted the CLI Kerberos authentication page.
-Cleaned up formatting.
-Fixed links.

## Documentation

( ) No documentation is needed.
(x) Sufficient documentation is included in this PR.
( ) Documentation PR is available with #prnumber.
( ) Documentation issue #issuenumber is filed, and can be handled later.

## Release notes

(x) No release notes entries required.
( ) Release notes entries required with the following suggested text:

```markdown
# Section
* Fix some things. ({issue}`issuenumber`)
```

### CODE REVIEW

File: docs/src/main/sphinx/installation/cli.rst [hunk 1]
```diff
@@ -314,6 +314,12 @@ To access a Trino cluster configured to use :doc:`/security/jwt`, use the
 Kerberos authentication
 ^^^^^^^^^^^^^^^^^^^^^^^
 
+The Trino :doc:`/installation/cli` can connect to a Trino cluster that has
+:doc:`/security/kerberos` enabled.
+
+Trino CLI execution
```
Comment (mosabua) [inline docs/src/main/sphinx/installation/cli.rst:320]: drop this title

File: docs/src/main/sphinx/installation/cli.rst [hunk 2]
```diff
@@ -314,6 +314,12 @@ To access a Trino cluster configured to use :doc:`/security/jwt`, use the
...
+~~~~~~~~~~~~~~~~~~~
+
 In addition to the options that are required when connecting to an unauthorized
```
Comment (mosabua) [inline docs/src/main/sphinx/installation/cli.rst:323]: too much "in addition" .. "additional" .. reword

File: docs/src/main/sphinx/installation/cli.rst [hunk 3]
```diff
@@ -358,8 +364,51 @@ The following table list the available options for Kerberos authentication:
   * - ``--krb5-service-principal-pattern``
     - Remote kerberos service principal pattern (default: ${SERVICE}@${HOST})
 
-See :doc:`/security/cli` for more information on configuring and using Kerberos
-with the CLI.
+Troubleshooting
+~~~~~~~~~~~~~~~
+
+Many of the same steps, that can be used when troubleshooting the :ref:`Trino
```
Comment (mosabua) [inline docs/src/main/sphinx/installation/cli.rst:370]: Where does this link to?

File: docs/src/main/sphinx/installation/cli.rst [hunk 4]
```diff
@@ -358,8 +364,51 @@ The following table list the available options for Kerberos authentication:
...
+coordinator <coordinator-troubleshooting>` apply to troubleshooting the CLI.
+
+Additional Kerberos debugging information
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+You can enable additional Kerberos debugging information for the Trino CLI
```
Comment (mosabua) [inline docs/src/main/sphinx/installation/cli.rst:376]: This approach needs to be broken up .. 

basically there are multiple topics here 

- add more JVM flags to run CLI 
- what are the JVM flags for Kerberos debugging
- example of using Kerberos options
please arrange a meeting so I can explain more in detai

Comment (mosabua) [reply docs/src/main/sphinx/installation/cli.rst:376]: and we need to cover these in separate sections in this doc

File: docs/src/main/sphinx/installation/cli.rst [hunk 5]
```diff
@@ -358,8 +364,51 @@ The following table list the available options for Kerberos authentication:
...
+      --krb5-config-path /etc/krb5.conf \
+      --krb5-principal someuser@EXAMPLE.COM \
+      --krb5-keytab-path /home/someuser/someuser.keytab \
+      --krb5-remote-service-name trino \
+      --keystore-path /tmp/trino.jks \
+      --keystore-password password \
+      --catalog <catalog> \
+      --schema <schema>
+
+The :ref:`additional resources <server_additional_resources>` listed in the
```
Comment (mosabua) [inline docs/src/main/sphinx/installation/cli.rst:402]: Need to check what we link to and reword

File: docs/src/main/sphinx/installation/cli.rst [hunk 6]
```diff
@@ -358,8 +364,51 @@ The following table list the available options for Kerberos authentication:
...
+documentation for setting up Kerberos authentication for the Trino coordinator
+may helpful when interpreting the Kerberos debugging messages.
+
+See `Troubleshooting Security
```
Comment (mosabua) [inline docs/src/main/sphinx/installation/cli.rst:406]: this should maybe be in the Kerberos authentication docs instead

File: docs/src/main/sphinx/security/kerberos.rst [hunk 7]
```diff
@@ -50,7 +50,8 @@ Configuration for TLS
 ^^^^^^^^^^^^^^^^^^^^^
 
 When using Kerberos authentication, access to the Trino coordinator must be
-through :doc:`TLS and HTTPS </security/tls>`.
+through HTTPS. If you have not yet configured TLS/HTTPS for your coordinator,
```
Comment (mosabua) [inline docs/src/main/sphinx/security/kerberos.rst:53]: undo this change

**Review by mosabua — CHANGES_REQUESTED** [review]
First round of input .. need to fix up a bunch of things

File: docs/src/main/sphinx/installation/cli.rst (hunk 5 above)
Comment (Jessie212) [reply docs/src/main/sphinx/installation/cli.rst:402]: This links to the debugging section in the Kerberos authentication doc.

File: docs/src/main/sphinx/installation/cli.rst (hunk 3 above)
Comment (Jessie212) [reply docs/src/main/sphinx/installation/cli.rst:370]: It links to the troubleshooting section of the Kerberos Authentication doc.

File: docs/src/main/sphinx/installation/cli.rst (hunk 4 above)
Comment (Jessie212) [reply docs/src/main/sphinx/installation/cli.rst:372]: Pushed my changes. Will do this in the next commit.

File: docs/src/main/sphinx/security/kerberos.rst [hunk 8]
```diff
@@ -50,7 +50,7 @@ Configuration for TLS
 ^^^^^^^^^^^^^^^^^^^^^
 
 When using Kerberos authentication, access to the Trino coordinator must be
-through :doc:`TLS and HTTPS </security/tls>`.
+through :doc:`HTTPS and TLS </security/tls>`.
```
Comment (Ordinant) [inline docs/src/main/sphinx/security/kerberos.rst:53]: I've tried to standardize on TLS first. There's even a Jira for that. 

TLS is the actual protocol used that we want people to get used to seeing.

File: docs/src/main/sphinx/installation/cli.rst [hunk 9]
```diff
@@ -314,10 +314,12 @@ To access a Trino cluster configured to use :doc:`/security/jwt`, use the
 Kerberos authentication
 ^^^^^^^^^^^^^^^^^^^^^^^
 
-In addition to the options that are required when connecting to an unauthorized
-Trino coordinator, invoking the CLI with Kerberos support enabled requires a
-number of additional command line options. The simplest way to invoke the CLI is
-with a wrapper script.
+The Trino :doc:`/installation/cli` can connect to a Trino cluster that has
```
Comment (Ordinant) [inline docs/src/main/sphinx/installation/cli.rst:317]: Aren't we IN that file now? So this is a circular cross-ref. It can just be text: `Trino CLI`.

File: docs/src/main/sphinx/installation/cli.rst [hunk 10]
```diff
@@ -334,7 +336,7 @@ with a wrapper script.
       --catalog <catalog> \
       --schema <schema>
 
-The following table list the available options for Kerberos authentication:
+The following table lists the available options for Kerberos authentication:
```
Comment (Ordinant) [inline docs/src/main/sphinx/installation/cli.rst:339]: You can possibly cross-ref `Kerberos authentication` to :doc:`/security/kerberos`, but that reference is already on line 318, so maybe this would be overkill? Get a consensus on tis.

File: docs/src/main/sphinx/installation/cli.rst [hunk 11]
```diff
@@ -358,8 +360,76 @@ The following table list the available options for Kerberos authentication:
   * - ``--krb5-service-principal-pattern``
     - Remote kerberos service principal pattern (default: ${SERVICE}@${HOST})
 
-See :doc:`/security/cli` for more information on configuring and using Kerberos
-with the CLI.
+Kerberos principals and keytab files
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Each user, who connects to the Trino coordinator, needs a Kerberos principal.
```
Comment (Ordinant) [inline docs/src/main/sphinx/installation/cli.rst:366]: Remove both commas.

File: docs/src/main/sphinx/installation/cli.rst [hunk 12]
```diff
@@ -358,8 +360,76 @@ The following table list the available options for Kerberos authentication:
...
+You need to create these users in Kerberos using `kadmin
```
Comment (Ordinant) [inline docs/src/main/sphinx/installation/cli.rst:367]: must instead of need to

File: docs/src/main/sphinx/installation/cli.rst [hunk 13]
```diff
@@ -358,8 +360,76 @@ The following table list the available options for Kerberos authentication:
...
+    > addprinc -randkey someuser@EXAMPLE.COM
+    > ktadd -k /home/someuser/someuser.keytab someuser@EXAMPLE.COM
+
+.. include:: ../security/ktadd-note.fragment
+
+Configuration for TLS
+~~~~~~~~~~~~~~~~~~~~~
+
+When using Kerberos authentication, access to the Trino coordinator must be
+through :doc:`HTTPS and TLS </security/tls>`.
```
Comment (Ordinant) [inline docs/src/main/sphinx/installation/cli.rst:387]: The title of the page you're linking to is ``TLS and HTTPS``.

File: docs/src/main/sphinx/installation/cli.rst [hunk 14]
```diff
@@ -358,8 +360,76 @@ The following table list the available options for Kerberos authentication:
...
+
+Troubleshooting
+~~~~~~~~~~~~~~~
+
+Many of the same steps, that can be used when troubleshooting the :ref:`Trino
```
Comment (Ordinant) [inline docs/src/main/sphinx/installation/cli.rst:392]: That comma is giving me hives. Remove it to save Uncle Barton.

File: docs/src/main/sphinx/installation/cli.rst [hunk 15]
```diff
@@ -358,8 +360,76 @@ The following table list the available options for Kerberos authentication:
...
+coordinator <coordinator-troubleshooting>` apply to troubleshooting the CLI.
+
+Additional Kerberos debugging information
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+You can enable additional Kerberos debugging information for the Trino CLI
+process by passing ``-Dsun.security.krb5.debug=true``,
+``-Dtrino.client.debugKerberos=true``, and
+``-Djava.security.debug=gssloginconfig,configfile,configparser,logincontext``
+as a JVM argument, when starting the CLI process. Doing so requires invoking
```
Comment (Ordinant) [inline docs/src/main/sphinx/installation/cli.rst:402]: Again, where do these random commas come from? Use a comma where you would take a breath when reading aloud. This one goes out.

**Review by Ordinant — COMMENTED** [review]
Light review, with comments left, knowing that this needs to be cut up later according to Manfred's comment.

File: docs/src/main/sphinx/installation/cli.rst (hunk 15 above)
Comment (Jessie212) [reply docs/src/main/sphinx/installation/cli.rst:402]: lol @Ordinant I don't know. Hasty copying and pasting and chopping, maybe. I'll slap a label on this PR, edit, and check links and such.

File: docs/src/main/sphinx/installation/cli.rst [hunk 16]
```diff
@@ -13,9 +13,10 @@ Requirements
 The CLI requires a Java virtual machine available on the path.
 It can be used with Java version 8 and higher.
 
-The CLI uses the :doc:`Trino client REST API
-</develop/client-protocol>` over HTTP/HTTPS to communicate with the
-coordinator on the cluster.
+The CLI uses the :doc:`Trino client REST API </develop/client-protocol>` over
```
Comment (mosabua) [inline docs/src/main/sphinx/installation/cli.rst:16]: Wrapping changes only here

File: docs/src/main/sphinx/installation/cli.rst [hunk 17]
```diff
@@ -328,27 +332,29 @@ To access a Trino cluster configured to use :doc:`/security/jwt`, use the
...
-In addition to the options that are required when connecting to an unauthorized
-Trino coordinator, invoking the CLI with Kerberos support enabled requires a
-number of additional command line options. The simplest way to invoke the CLI is
-with a wrapper script.
+The Trino CLI can connect to a Trino cluster that has :doc:`/security/kerberos`
+enabled.
+
+Invoking the CLI with Kerberos support enabled requires a number of additional
+command line options. You also need the :ref:`Kerberos configuration files
+<server_kerberos_principals>` for your user on the maching running the CLI. The
```
Comment (mosabua) [inline docs/src/main/sphinx/installation/cli.rst:340]: machine

File: docs/src/main/sphinx/installation/cli.rst [hunk 18]
```diff
@@ -328,27 +332,29 @@ To access a Trino cluster configured to use :doc:`/security/jwt`, use the
...
+simplest way to invoke the CLI is with a wrapper script.
```
Comment (mosabua) [inline docs/src/main/sphinx/installation/cli.rst:341]: script:

File: docs/src/main/sphinx/security/kerberos.rst [hunk 19]
```diff
@@ -29,12 +29,13 @@ Kerberos principals and keytab files
 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 
 The Trino coordinator needs a Kerberos principal, as do users who are going to
-connect to the Trino coordinator. You need to create these users in
-Kerberos using `kadmin
+connect to the Trino coordinator. You need to create these users in Kerberos
```
Comment (mosabua) [inline docs/src/main/sphinx/security/kerberos.rst:32]: Wrapping fix only

File: docs/src/main/sphinx/security/kerberos.rst [hunk 20]
```diff
@@ -29,12 +29,13 @@ Kerberos principals and keytab files
...
+using `kadmin
 <http://web.mit.edu/kerberos/krb5-latest/doc/admin/admin_commands/kadmin_local.html>`_.
 
 In addition, the Trino coordinator needs a `keytab file
-<http://web.mit.edu/kerberos/krb5-devel/doc/basic/keytab_def.html>`_. After you create the principal, you can create the keytab file using :command:`kadmin`
+<http://web.mit.edu/kerberos/krb5-devel/doc/basic/keytab_def.html>`_. After you
```
Comment (mosabua) [inline docs/src/main/sphinx/security/kerberos.rst:37]: Wrapping fix only

File: docs/src/main/sphinx/security/kerberos.rst [hunk 21]
```diff
@@ -56,8 +57,8 @@ System access control plugin
 ----------------------------
 
 A Trino coordinator with Kerberos enabled probably needs a
-:doc:`/develop/system-access-control` plugin to achieve
-the desired level of security.
+:doc:`/develop/system-access-control` plugin to achieve the desired level of
```
Comment (mosabua) [inline docs/src/main/sphinx/security/kerberos.rst:60]: Wrapping fix only

File: docs/src/main/sphinx/security/kerberos.rst [hunk 22]
```diff
@@ -77,7 +78,8 @@ config.properties
 ^^^^^^^^^^^^^^^^^
 
 Kerberos authentication is configured in the coordinator node's
-:file:`config.properties` file. The entries that need to be added are listed below.
+:file:`config.properties` file. The entries that need to be added are listed
```
Comment (mosabua) [inline docs/src/main/sphinx/security/kerberos.rst:81]: Wrapping fix only

File: docs/src/main/sphinx/security/kerberos.rst [hunk 23]
```diff
@@ -135,35 +137,38 @@ access-controls.properties
 ^^^^^^^^^^^^^^^^^^^^^^^^^^
 
 At a minimum, an :file:`access-control.properties` file must contain an
-``access-control.name`` property.  All other configuration is specific
-for the implementation being configured.
-See :doc:`/develop/system-access-control` for details.
+``access-control.name`` property.  All other configuration is specific for the
+implementation being configured. See :doc:`/develop/system-access-control` for
```
Comment (mosabua) [inline docs/src/main/sphinx/security/kerberos.rst:141]: Wrapping fix only

File: docs/src/main/sphinx/security/kerberos.rst [hunk 24]
```diff
@@ -135,35 +137,38 @@ access-controls.properties
...
 User mapping
 ------------
 
-After authenticating with Kerberos, the Trino server receives the user's principal which is typically similar to
-an email address.  For example, when ``alice`` logs in Trino might receive ``alice@example.com``.  By default,
-Trino will use the full Kerberos principal name, but this can be mapped to a shorter name using a user-mapping
-pattern.  For simple mapping rules, the  ``http-server.authentication.krb5.user-mapping.pattern`` configuration
-property can be set to a Java regular expression, and Trino will use the value of the first matcher group.  If the
-regular expression does not match, the authentication is denied.  For more complex user-mapping rules, see
+After authenticating with Kerberos, the Trino server receives the user's
```
Comment (mosabua) [inline docs/src/main/sphinx/security/kerberos.rst:149]: Wrapping fix only

**Review by mosabua — CHANGES_REQUESTED** [review]
Minor nits

File: docs/src/main/sphinx/installation/cli.rst (hunk 17 above)
Comment (Jessie212) [reply docs/src/main/sphinx/installation/cli.rst]: This was corrected. Forgot to push the changes.

**Review by mosabua — APPROVED** [review]
Nice clean up.

File: docs/src/main/sphinx/security/kerberos.rst [hunk 25]
```diff
@@ -135,35 +137,38 @@ access-controls.properties
...
+principal which is typically similar to an email address. For example, when
+``alice`` logs in Trino might receive ``alice@example.com``. By default, Trino
+will use the full Kerberos principal name, but this can be mapped to a shorter
```
Comment (Ordinant) [inline docs/src/main/sphinx/security/kerberos.rst:152]: Trino uses (no future tense)

File: docs/src/main/sphinx/security/kerberos.rst [hunk 26]
```diff
@@ -135,35 +137,38 @@ access-controls.properties
...
+name using a user-mapping pattern. For simple mapping rules, the
+``http-server.authentication.krb5.user-mapping.pattern`` configuration property
+can be set to a Java regular expression, and Trino will use the value of the
```
Comment (Ordinant) [inline docs/src/main/sphinx/security/kerberos.rst:155]: Trino uses
(the future doesn't exist for tech writers)

File: docs/src/main/sphinx/security/kerberos.rst [hunk 27]
```diff
@@ -135,35 +137,38 @@ access-controls.properties
...
-independently verify some of the configuration outside of Trino, to help narrow
+independently verify some of the configuration outside of Trino to help narrow
 your focus when trying to solve a problem.
 
 Kerberos verification
 ^^^^^^^^^^^^^^^^^^^^^
 
 Ensure that you can connect to the KDC from the Trino coordinator using
-:command:`telnet`.
+:command:`telnet`:
```
Comment (Ordinant) [inline docs/src/main/sphinx/security/kerberos.rst:171]: The insecure `telnet` command is no longer included with macOS or with modern Linuxes. Shouldn't this be `ssh` nowadays? @mosabua

File: docs/src/main/sphinx/installation/cli.rst [hunk 28]
```diff
@@ -37,7 +38,10 @@ the version:
 
     java -jar trino-cli-*-executable.jar --version
 
-The syntax can be used for the examples in the following sections.
+The syntax can be used for the examples in the following sections. In addition,
+using the ``java`` command allows you to add configuration options for the Java
+runtime with the ``-D`` syntax. You can use this for debugging and
+troubleshooting, for example :ref:`cli-kerberos-debug`.
```
Comment (Ordinant) [inline docs/src/main/sphinx/installation/cli.rst:44]: Two things here. 
1. Replace "for example" with "such as for".  But see point 2 first.
"for example" is not a conjunction like "but" that can join two parts of a sentence. It should always begin a sentence or clause.
2. As I remember, this :ref: form will take the wording of the following title and place it here. That will place an odd init-cap "Additional" in the middle of a sentence here. Should this be something like:

troubleshooting, such as when :ref:` specifying additional kerberos debug options <cli-kerberos-debug`.

(I haven't built this doc to see whether I got that syntax right.)

**Review by Ordinant — COMMENTED** [review]
Overall, it looks great. But I still have quibbles.

**Review by mosabua — COMMENTED** [review]
All good now. Thank you. Please merge @martint or @electrum

**Review by Ordinant — APPROVED** [review]
Ship it.

**Comment by Jessie212** [comment]
@electrum and @martint Can you take a look at this PR and merge?

**Review by electrum — APPROVED** [review]

File: docs/src/main/sphinx/client/cli.rst [hunk 29]
```diff
@@ -373,8 +380,34 @@ The following table list the available options for Kerberos authentication:
...
+
+You can enable additional Kerberos debugging information for the Trino CLI
+process by passing ``-Dsun.security.krb5.debug=true``,
+``-Dtrino.client.debugKerberos=true``, and
+``-Djava.security.debug=gssloginconfig,configfile,configparser,logincontext``
+as a JVM argument when :ref:`starting the CLI process <cli-installation>`:
+
+.. code-block:: text
+
+    #!/bin/bash
```
Comment (electrum) [inline docs/src/main/sphinx/client/cli.rst:396]: Remove this, since we're showing how to run the command, rather than creating a shell script in a file.

**Comment by mosabua** [comment]
Thank you @Jessie212 and @electrum
